### DISCUSSION TITLE
Enable View and View Model in same file

### DISCUSSION THREAD

**Discussion by johanpolson** [AvaloniaUI/Avalonia#15200] [Ideas]
Often ViewModels are small and only build for one View, in these cases it would be nice to be able to have view and View Model in the same file.

## Example
``` xmal
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:vm="clr-namespace:Example.ViewModels"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="Example.Views.MainView"
             x:DataType="vm:MainViewModel">
  <Design.DataContext>
    <vm:MainViewModel />
  </Design.DataContext>

  <ContentPresenter Content="{Binding Page}" />
</UserControl>
<Code><![CDATA[

namespace Example.ViewModels;

public partial class MainViewModel : ViewModelBase
{
    private ViewModelBase? m_page;

    public ViewModelBase? Page { get => m_page; set => this.RaiseAndSetIfChanged(ref m_page, value); }

    public MainViewModel() {
    }
}

]]><Code>
```

**Comment by timunie** [comment]
not possible in XAML however one can do that in pure C# as you can also easily bind from code behind. For example: https://github.com/AvaloniaUI/Avalonia.Markup.Declarative

**Comment by thevortexcloud** [comment]
I am not sure why you would want this when there is already a code behind file that can be used for this if you really want to do it.

**Comment by stevemonaco** [comment]
I suspect it would require a lot of IDE work to get the C# tooling to work within a XAML context for autocompletion.

**Reply by PhilWeisz** [reply]
that is a good question,... has anyone got a link for comparing with anything similiar?

**Reply by thevortexcloud** [reply]
The closest thing I know of is the Razor syntax which is fairly special. It lets you more or less seamlessly mix C# and HTML at the same time. EG this is valid Razor:

```razor
<body>
@{ var i = 0; }
@while (i < people.Length)
{
    var person = people[i];
    <p>Name: @person.Name</p>
    <p>Age: @person.Age</p>

    i++;
}
</body>

```

Which will output all the paragraph tags to the page based on how many people are in the people list.

https://learn.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-8.0

So there is precedent for this sort of thing. However, IIRC Microsoft themselves have talked about how complicated it was to add support for Visual Studio for Razor as it's more or less two different lexers smooshed together with a bunch of weird rules.

The proposal from OP is a lot simpler than Razor, but it's probably still fairly complicated.

**Comment by PhilWeisz** [comment]
> Often ViewModels are small and only build for one View, in these cases it would be nice to be able to have view and > View Model in the same file
 
Yes CDATA ... should be the correct syntax actually.
Although this req somewhat violates best practices of coding in XML-based-UI-Design-languages frameworks  
I rather tend to agree, because  the syntax is long enough already, bogus files on top

I guess you can do some trickery in the .csproj file - if you use the pre-build phase you can modify your code and files
as you want to, as long as the compiler has a working source at buildtime

 Before the actual build:

- Match  between the cdata- tags to get the inlined .cs-code (<!CDATA[.... ])
- Cut it out
- Create a new .cs file and paste it inside
- You need to execute the add source to solution command (VS-api)
- Build

You could use .csproj's-capabilities, and run code in the pre-build-phase

 _Before the actual  build..:_

**<UserControl> **
**</UserControl>**
 **  <![CDATA[ ......X.....> **
 
In fact there are several different ways to invoke code to modify your repository (generated):

<details>
<summary>Other Options </summary
MSBuild Tasks: MSBuild, the build platform for .NET, allows the execution of custom tasks. You can define tasks in your .csproj or .targets files that can execute before the build process. These tasks can be written in C# or any other .NET-compatible language and can perform a wide range of operations like file manipulation, code generation, or even invoking external tools and scripts.

Pre-build Events: Visual Studio and MSBuild allow you to specify pre-build events in your project properties. These are essentially command line instructions that run before your project is built. Pre-build events can call batch files, execute PowerShell scripts, or run any executable with command line arguments.

Source Generators in C# 9 and above: Starting with C# 9, Source Generators allow you to generate additional C# source files automatically as part of the compilation process. This is a powerful feature for metaprogramming where code can generate other code.

T4 Templates: T4 (Text Template Transformation Toolkit) templates are a part of Visual Studio and can be used to generate C# code, configuration files, or any other type of text file. T4 templates can be set to transform on each build automatically, providing a powerful way to produce source code or other files dynamically before the build process.

Roslyn Analyzers: While primarily used for code analysis and providing real-time feedback in the IDE, Roslyn analyzers can also be used to perform code modifications or injections before the build, though this is a more advanced and less common use case.

Custom Build Tools: For scenarios where built-in mechanisms aren't sufficient, you can develop custom tools or scripts in the language of your choice (Python, Ruby, Bash, etc.) and invoke them during the pre-build phase using any of the aforementioned mechanisms. This could include anything from code linters to more complex code transformation tools.</summary>
</details >

**Comment by mathmakesmemighty** [comment]
I see. However... it mixes 2 new "technologies"  that have different resolution-targets.(axaml is similiar to Wpf's xaml. We should  maybe do both, checking the Xaml-AUTOcomplete official/legacy material and the current code/behavior in avaloniaOn 5 Apr 2024 04:23, thevortexcloud ***@***.***> wrote:
The closest thing I know of is the Razor syntax which is fairly special. It lets you more or less seamlessly mix C# and HTML at the same time. EG this is valid Razor:
@{ var i = 0; }
@while (i < people.Length)
{
    var person = people[i];
    <p>Name: @person.Name</p>
    <p>Age: @person.Age</p>

    i++;
}

Which will output all the paragraph tags to the page based on how many people are in the people list.
https://learn.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-8.0
So there is precedent for this sort of thing, but it's fairly complicated. IIRC Microsoft themselves have talked about how complicated it was to add support for Visual Studio.

**Reply by thevortexcloud** [reply]
The point is that adding support for this is:

1. A lot of work (not just for Avalonia, but also IDE vendors)
2. Unlikely to get widespread use outside of quick prototyping scenarios
3. Does not really solve any problems that could not already be solved by just using the existing code behind file (there is nothing stopping you from declaring your VM there)
4. Encourages bad patterns
5. Sort of breaks MVVM by directly coupling the view file to the view model
